server/src/main/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregationBuilder.java

```diff
@@ -39,21 +40,56 @@ public class ScriptedMetricAggregationBuilder extends AbstractAggregationBuilder
     private static final ParseField REDUCE_SCRIPT_FIELD = new ParseField("reduce_script");
     private static final ParseField PARAMS_FIELD = new ParseField("params");
 
-    public static final ConstructingObjectParser<ScriptedMetricAggregationBuilder, String> PARSER = new ConstructingObjectParser<>(
-        NAME,
-        false,
-        (args, name) -> {
-            ScriptedMetricAggregationBuilder builder = new ScriptedMetricAggregationBuilder(name);
-            builder.mapScript((Script) args[0]);
-            return builder;
+    private static void validateScript(String scriptName, String aggName, Script script, List<String> allowedScripts) {
+        if (script == null) {
+            throw new IllegalArgumentException("[" + scriptName + "] must not be null: [" + aggName + "]");
+        }
+        if (allowedScripts.isEmpty() == false) {
+            if (allowedScripts.contains(script.getIdOrCode()) == false) {
+                throw new IllegalArgumentException("[" + scriptName + "] contains not allowed script: [" + aggName + "]");
+            }
+        }
+    }
```

Comment:
I think i'd prefer to put the allowlist in `ScriptedMetricAggregationBuilder#doBuild`. There's an `AggregationContext` which you can wire the validation logic into. When we build `ProductionAggregationContext` we already have a copy of `ClusterSettings` which I *think* is where we'd grab the list.

I feel like parsing is too early. It's not actually too early because we don't have any way around the public API, but I still think of our REST layer as a wrapper around the "real" ES so I don't want to put the security bits in it.

Reply:
Thanks! You're right, this is better.